control-plane/api/v1alpha1/jwtprovider_types.go

```diff
@@ -0,0 +1,663 @@
+	// means with which to fetch a key set from a remote server.
+	JSONWebKeySet *JSONWebKeySet `json:"jsonWebKeySet,omitempty"`
+
+	// Issuer is the entity that must have issued the JWT.
+	// This value must match the "iss" claim of the token.
+	Issuer string `json:"issuer,omitempty"`
+
+	// Audiences is the set of audiences the JWT is allowed to access.
+	// If specified, all JWTs verified with this provider must address
+	// at least one of these to be considered valid.
+	Audiences []string `json:"audiences,omitempty"`
+
+	// Locations where the JWT will be present in requests.
+	// Envoy will check all of these locations to extract a JWT.
+	// If no locations are specified Envoy will default to:
+	// 1. Authorization header with Bearer schema:
+	//    "Authorization: Bearer <token>"
+	// 2. accessToken query parameter.
+	Locations []*JWTLocation `json:"locations,omitempty"`
+
+	// Forwarding defines rules for forwarding verified JWTs to the backend.
+	Forwarding *JWTForwardingConfig `json:"forwarding,omitempty"`
+
+	// ClockSkewSeconds specifies the maximum allowable time difference
+	// from clock skew when validating the "exp" (Expiration) and "nbf"
+	// (Not Before) claims.
+	//
+	// Default value is 30 seconds.
+	ClockSkewSeconds int `json:"clockSkewSeconds,omitempty"`
+
+	// CacheConfig defines configuration for caching the validation
+	// result for previously seen JWTs. Caching results can speed up
+	// verification when individual tokens are expected to be handled
+	// multiple times.
+	CacheConfig *JWTCacheConfig `json:"cacheConfig,omitempty"`
+}
+
+type JWTLocations []*JWTLocation
+
+func (j JWTLocations) toConsul() []*capi.JWTLocation {
+	if j == nil {
+		return nil
+	}
+	result := make([]*capi.JWTLocation, 0, len(j))
+	for _, loc := range j {
+		result = append(result, loc.toConsul())
+	}
+	return result
+}
+
+func (j JWTLocations) validate(path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	for i, loc := range j {
+		errs = append(errs, loc.validate(path.Index(i))...)
+	}
+	return errs
+}
+
+// JWTLocation is a location where the JWT could be present in requests.
+//
+// Only one of Header, QueryParam, or Cookie can be specified.
+type JWTLocation struct {
+	// Header defines how to extract a JWT from an HTTP request header.
+	Header *JWTLocationHeader `json:"header,omitempty"`
+
+	// QueryParam defines how to extract a JWT from an HTTP request
+	// query parameter.
+	QueryParam *JWTLocationQueryParam `json:"queryParam,omitempty"`
+
+	// Cookie defines how to extract a JWT from an HTTP request cookie.
+	Cookie *JWTLocationCookie `json:"cookie,omitempty"`
+}
+
+func (j *JWTLocation) toConsul() *capi.JWTLocation {
+	if j == nil {
+		return nil
+	}
+	return &capi.JWTLocation{
+		Header:     j.Header.toConsul(),
+		QueryParam: j.QueryParam.toConsul(),
+		Cookie:     j.Cookie.toConsul(),
+	}
+}
+
+func (j *JWTLocation) validate(path *field.Path) field.ErrorList {
```

Comment:
Throughout this file, I tried to match the validations we implemented in https://github.com/hashicorp/consul/blob/main/agent/structs/config_entry_jwt_provider.go